fdbserver/BlobManager.actor.cpp

```diff
@@ -1314,6 +1320,585 @@ ACTOR Future<Void> blobWorkerRecruiter(
+	Version latestSnapshotVersion = invalidVersion;
+
+	state std::vector<Future<Void>> deletions; // deletion work per file
+	state std::vector<Key> deletedFileKeys; // keys for deleted files
+	state std::vector<std::string> filesToDelete; // TODO: remove evenutally, just for debugging
+
+	// TODO: binary search these snapshot files for latestSnapshotVersion
+	for (int idx = files.snapshotFiles.size() - 1; idx >= 0; --idx) {
+		// if we already found the latestSnapshotVersion, this snapshot can be deleted
+		if (latestSnapshotVersion != invalidVersion) {
+			std::string fname = files.snapshotFiles[idx].filename;
+			deletions.emplace_back(self->bstore->deleteFile(fname));
+			deletedFileKeys.emplace_back(blobGranuleFileKeyFor(granuleId, 'S', files.snapshotFiles[idx].version));
+			filesToDelete.emplace_back(fname);
+		} else if (files.snapshotFiles[idx].version <= pruneVersion) {
+			// otherwise if this is the FIRST snapshot file with version < pruneVersion,
+			// then we found our latestSnapshotVersion (FIRST since we are traversing in reverse)
+			latestSnapshotVersion = files.snapshotFiles[idx].version;
+		}
+	}
+
+	// we would have only partially deleted the granule if such a snapshot existed
+	ASSERT(latestSnapshotVersion != invalidVersion);
+
+	// delete all delta files older than latestSnapshotVersion
+	for (auto deltaFile : files.deltaFiles) {
+		// traversing in fwd direction, so stop once we find the first delta file past the latestSnapshotVersion
+		if (deltaFile.version > latestSnapshotVersion) {
+			break;
+		}
+
+		// otherwise deltaFile.version <= latestSnapshotVersion so delete it
+		// == should also be deleted because the last delta file before a snapshot would have the same version
+		std::string fname = deltaFile.filename;
+		deletions.emplace_back(self->bstore->deleteFile(fname));
+		deletedFileKeys.emplace_back(blobGranuleFileKeyFor(granuleId, 'D', deltaFile.version));
+		filesToDelete.emplace_back(fname);
+	}
+
+	if (BM_DEBUG) {
+		printf("Partially deleting granule %s: deleting %d files\n", granuleId.toString().c_str(), deletions.size());
+		for (auto filename : filesToDelete) {
+			printf(" - %s\n", filename.c_str());
+		}
+	}
+
+	// TODO: the following comment relies on the assumption that BWs will not get requests to
+	// read data that was already pruned. confirm assumption is fine. otherwise, we'd need
+	// to communicate with BWs here and have them ack the pruneVersion
+
+	// delete the files before the corresponding metadata.
+	// this could lead to dangling pointers in fdb, but we should never read data older than
+	// pruneVersion anyways, and we can clean up the keys the next time around.
+	// deleting files before corresponding metadata reduces the # of orphaned files.
+	wait(waitForAll(deletions));
+
+	// delete metadata in FDB (deleted file keys)
+	// TODO: do we need to also update the start version for the history entry?
```

Comment:
To confirm

Reply:
Yeah I think so. Right now we assume in a couple places that the first snapshot file's version == the granule start version in the history entry

Reply:
Discussed offline. We will need avoid using this assumption - otherwise, we'd need to update the child's `parents` field with the new parent start version